File: Assignment 5: Kd-Trees/Python/PointSET.py
from matplotlib import pyplot as plt
from Point2D import Point2D
from RectHV import RectHV

class PointSET:
    """
    A class representing a set of 2-dimensional points.

    Expected Performance:
    - Insertion: O(1) on average, as points are appended to a list.
    - Contains: O(N) in the worst case, where N is the number of points, since it requires a linear search.
    - Nearest Neighbor Search: O(N) in the worst case, as it involves checking the distance to each point.
    - Range Search: O(N) in the worst case, as it involves checking each point to see if it lies within the given rectangle.
    """

    def __init__(self):
        """
        Initialize an empty set of points.
        """
        self.set = []

    def isEmpty(self):
        """
        Check if the set of points is empty.

        Returns:
            bool: True if the set is empty, False otherwise.
        """
        return not self.set

    def size(self):
        """
        Get the number of points in the set.

        Returns:
            int: The number of points in the set.
        """
        return len(self.set)

    def insert(self, point):
        """
        Insert a new point into the set.

        Args:
            point (Point2D): The 2-dimensional point to be inserted.

        Raises:
            ValueError: If the point is None.
        """
        if point is None:
            raise ValueError("Invalid argument")

        self.set.append(point)

    def get(self, point):
        """
        Retrieve a point from the set.

        Args:
            point (Point2D): The 2-dimensional point to be retrieved.

        Raises:
            ValueError: If the point is None.

        Returns:
            Point2D: The point if found, None otherwise.
        """
        if point is None:
            raise ValueError("Invalid argument")

        if point in self.set:
            return point

        return None

    def contains(self, point):
        """
        Check if a point exists in the set.

        Args:
            point (Point2D): The 2-dimensional point to be checked.

        Raises:
            ValueError: If the point is None.

        Returns:
            bool: True if the point exists in the set, False otherwise.
        """
        if point is None:
            raise ValueError("Invalid argument")

        return point in self.set

    def draw(self):
        """
        Draw all points in the set.

        Sets up the plot limits and draws each point.
        """
        plt.xlim(0, 1)
        plt.ylim(0, 1)

        for point in self.set:
            point.draw()

        plt.show()

    def range(self, rect):
        """
        Find all points in the set that lie within a given rectangle.

        Args:
            rect (RectHV): The rectangle to search within.

        Raises:
            ValueError: If the rectangle is None.

        Returns:
            list: A list of points within the given rectangle.
        """
        if rect is None:
            raise ValueError("Invalid argument")

        insideRec = []

        for point in self.set:
            if rect.contains(point):
                insideRec.append(point)

        return insideRec

    def nearest(self, point):
        """
        Find the nearest neighbor to a given point in the set.

        Args:
            point (Point2D): The 2-dimensional point to find the nearest neighbor for.

        Raises:
            ValueError: If the point is None.

        Returns:
            Point2D: The nearest point in the set to the given point.
        """
        if point is None:
            raise ValueError("Invalid argument")

        if self.isEmpty():
            return None

        champion = self.set[0]
        
        for p in self.set:
            if p.distanceSquaredTo(point) < champion.distanceSquaredTo(point):
                champion = p
        return champion



# Example usage
if __name__ == "__main__":
    # Create an instance of PointSET
    pointSet = PointSET()

    # Insert points
    points = [
    Point2D(0.1, 0.3), Point2D(0.5, 0.6), Point2D(0.8, 0.9), Point2D(0.4, 0.2),
    Point2D(0.1, 0.4), Point2D(0.7, 0.4), Point2D(0.3, 0.8), Point2D(0.6, 0.1),
    Point2D(0.9, 0.7), Point2D(0.2, 0.5), Point2D(0.5, 0.3), Point2D(0.8, 0.2)]

    for point in points:
        pointSet.insert(point)

    # Check number of elements in the tree
    print("Number of elements in the tree = ", pointSet.size())

    # Check if the tree contains a point
    p = Point2D(0.7, 0.2)
    print(f"Does the tree countain the point {p}? {pointSet.contains(p)}")

    # Check nearest point
    point = Point2D(0.3, 0.35)
    print(f"Closest point to {point}: {pointSet.nearest(point)}")

    # Check points inside a rectangle
    rect = RectHV(0.15, 0.15, 0.75, 0.75)
    pointsInRect = pointSet.range(rect)
    print("Points inside the rectangle:", *[p for p in sorted(pointsInRect)])

    # Draw the elements
    rect.draw()
    point.draw('red')
    pointSet.draw()

File: Assignment 5 - Kd-Trees/Java/PointSET.java
import edu.princeton.cs.algs4.Point2D;
import edu.princeton.cs.algs4.RectHV;
import edu.princeton.cs.algs4.StdDraw;
import edu.princeton.cs.algs4.StdOut;

import java.util.LinkedList;
import java.util.TreeSet;

public class PointSET {
    private TreeSet<Point2D> treeset;

    /**
     * Constructs an empty set of points.
     */
    public PointSET() {
        treeset = new TreeSet<>();
    }

    /**
     * Checks if the set is empty.
     *
     * @return {@code true} if the set is empty, {@code false} otherwise
     */
    public boolean isEmpty() {
        return treeset.isEmpty();
    }

    /**
     * Returns the number of points in the set.
     *
     * @return the number of points in the set
     */
    public int size() {
        return treeset.size();
    }

    /**
     * Adds the point to the set (if it is not already in the set).
     *
     * @param p the point to be added
     * @throws IllegalArgumentException if the point is null
     */
    public void insert(Point2D p) {
        if (p == null) throw new IllegalArgumentException();
        treeset.add(p);
    }

    /**
     * Checks if the set contains the specified point.
     *
     * @param p the point to be checked
     * @return {@code true} if the set contains the point, {@code false} otherwise
     * @throws IllegalArgumentException if the point is null
     */
    public boolean contains(Point2D p) {
        if (p == null) throw new IllegalArgumentException();
        return treeset.contains(p);
    }

    /**
     * Draws all points to standard draw.
     */
    public void draw() {
        StdDraw.setXscale(0, 1);
        StdDraw.setYscale(0, 1);
        // StdDraw.setPenRadius(0.01);
        StdDraw.setPenColor(StdDraw.BLACK);
        for (Point2D p : treeset) {
            p.draw();
        }
    }

    /**
     * Returns all points that are inside the rectangle (or on the boundary).
     *
     * @param rect the rectangle
     * @return an iterable of points inside the rectangle
     * @throws IllegalArgumentException if the rectangle is null
     */
    public Iterable<Point2D> range(RectHV rect) {
        if (rect == null) throw new IllegalArgumentException();
        LinkedList<Point2D> insideRect = new LinkedList<>();
        for (Point2D p : treeset) {
            if (rect.contains(p)) {
                insideRect.add(p);
            }
        }
        return insideRect;
    }

    /**
     * Finds a nearest neighbor in the set to point p; returns null if the set is empty.
     *
     * @param p the point to find the nearest neighbor to
     * @return the nearest neighbor in the set to point p, or null if the set is empty
     * @throws IllegalArgumentException if the point is null
     */
    public Point2D nearest(Point2D p) {
        if (p == null) throw new IllegalArgumentException();
        if (treeset.isEmpty()) return null;
        Point2D nearest = treeset.first();
        for (Point2D point : treeset) {
            if (p.distanceSquaredTo(point) < p.distanceSquaredTo(nearest)) {
                nearest = point;
            }
        }
        return nearest;
    }

    /**
     * Unit testing of the methods (optional).
     *
     * @param args the command-line arguments
     */
    public static void main(String[] args) {
        PointSET set = new PointSET();

        // Insert points
        set.insert(new Point2D(0.7, 0.2));
        set.insert(new Point2D(0.5, 0.4));
        set.insert(new Point2D(0.2, 0.3));
        set.insert(new Point2D(0.4, 0.7));
        set.insert(new Point2D(0.9, 0.6));

        // Create a rectangle
        RectHV rect = new RectHV(0.15, 0.25, 0.5, 0.45);
        StdDraw.setPenColor(StdDraw.RED);
        rect.draw();

        // Draw all points in the set
        set.draw();

        // Print points inside the rectangle
        StdOut.println("Points inside the rectangle");
        for (Point2D p : set.range(rect)) {
            StdOut.println(p);
        }

        // Find and print the closest point to (0, 0)
        StdOut.println("Closest point to (0, 0):");
        Point2D closest = set.nearest(new Point2D(0, 0));
        StdOut.println(closest);
    }
}